### ISSUE TITLE
Failed to load firmware: 'NoneType' has no attribute 'join'

### ISSUE THREAD

**Issue by fredzyda** [autosportlabs/RaceCapture_App#1310]
When attempting to load firmware to my RCP Pro Mk2 from my mac, I keep getting the above error message. This happens on version 1.8.0 and 1.7.6. I've included a screen shot of the error message. The progress gauge thing never gets above 0, it crashes after "processing" and after the RCP has rebooted.

<img width="1024" alt="firmwarefailure" src="https://cloud.githubusercontent.com/assets/1243157/20822391/28dd0bd6-b7ff-11e6-851d-59641e85b8fa.png">

I ran the application in a terminal and got the following output while trying to do the firmware update starting from an already booted and connected RCP:

```INFO   ] [FirmwareUpdateView] Disabling RaceCapture Communcications
[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[INFO   ] [TelemetryManager] on_data_connected: False
[INFO   ] [DataBusPump ] sample_worker exiting
[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[INFO   ] [FwUpdater   ] Searching for an active bootloader
[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[INFO   ] [FirmwareUpdateView] Re-enabling RaceCapture Communications
 Exception ValueError: 'could not convert string to float: ' in 'kivy.properties.dpi2px' ignored
[INFO   ] [RCPAPI      ] trying last known device first: /dev/cu.usbmodem1421
[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[INFO   ] [RCPAPI      ] trying last known device first: /dev/cu.usbmodem1421
[ERROR  ] [RCPAPI      ] sendCommand exception caused by 'Port Closed'
[ERROR  ] [Traceback (most recent call last)]
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/racecapture/api/rcpapi.py", line 429, in sendCommand
  File "/Users/rdoherty/Documents/projects/RaceCapture_App/install/osx/RaceCapture.app/Contents/Resources/yourapp/autosportlabs/comms/comms.py", line 173, in write_message
PortNotOpenException: caused by 'Port Closed'

[INFO   ] [RCPAPI      ] trying last known device first: /dev/cu.usbmodem1421

```

If I boot the RCP while holding down the button, then try flashing the firmware, I get the following (slightly different) error output:

```[INFO   ] [FirmwareUpdateView] Disabling RaceCapture Communcications
[INFO   ] [FwUpdater   ] Searching for an active bootloader
[INFO   ] [FirmwareUpdateView] Re-enabling RaceCapture Communications
 Exception ValueError: 'could not convert string to float: ' in 'kivy.properties.dpi2px' ignored
[INFO   ] [RCPAPI      ] trying last known device first: /dev/cu.usbmodem1421
```

The output to the GUI is the same in either case.

I would be happy to get better logs to you if you tell me what I need.

**Comment by ryandoherty** [comment]
Hi @fredzyda, thanks for filing a ticket. What version of the app are you using and version of OSX? Does the app detect RCP before you start the firmware upgrade process? Thanks!

**Comment by brentpicasso** [comment]
Hi @fredzyda  - also, to narrow it down, are you able to compare behavior on Windows? Just want to narrow down the source of the issue.

**Comment by fredzyda** [comment]
Hi @ryandoherty, I tried versions 1.8.0 and 1.7.6 and got the same behavior in both versions. Is there another version I should try? @brentpicasso, I don't have access to a windows box at the moment, sadly.

**Comment by mattlqx** [comment]
I'm having this same issue on RaceCapture 2.10.1 trying to update firmware to 2.10.2 on OS X 10.12.3. The app does see the RC without issue.

**Comment by fredzyda** [comment]
I am still having this same problem, even with a fresh install of the RCP software on my mac...

**Comment by fredzyda** [comment]
@brentpicasso, I finally tried a windows box and the firmware update worked just fine. It appears this may be an OS X specific issue. I am currently on OS X version 10.12.3 running RaceCapture version 1.8.0.

**Comment by brentpicasso** [comment]
Thanks, fredzyda - because we can't reproduce it locally we'll give it a shot with where the problem might be happening.   @mattlqx - are you able to catch this in the act with a stack trace?  A hint might be it failing right after "Re-enabling RaceCapture Communications" in the log dump above. 

-Brent

**Comment by mattlqx** [comment]
Hooked up pdb and this is what I've got... not sure if this is exactly what you're looking for though.

```
(Pdb) n
> /Users/matt/Documents/Code/RaceCapture_App/autosportlabs/racecapture/views/configuration/rcp/firmwareupdateview.py(142)_update_thread()
-> port = fu.scan_for_device()
(Pdb) n
AttributeError: Attribut...'join'",)
> /Users/matt/Documents/Code/RaceCapture_App/autosportlabs/racecapture/views/configuration/rcp/firmwareupdateview.py(142)_update_thread()
-> port = fu.scan_for_device()
(Pdb) bt
  /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py(783)__bootstrap()
-> self.__bootstrap_inner()
  /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py(810)__bootstrap_inner()
-> self.run()
  /System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py(763)run()
-> self.__target(*self.__args, **self.__kwargs)
> /Users/matt/Documents/Code/RaceCapture_App/autosportlabs/racecapture/views/configuration/rcp/firmwareupdateview.py(142)_update_thread()
-> port = fu.scan_for_device()
  /Applications/Kivy.app/Contents/Resources/venv/lib/python2.7/site-packages/asl_f4_loader/fw_update.py(107)scan_for_device()
-> rsp = self._send_ping(p)
  /Applications/Kivy.app/Contents/Resources/venv/lib/python2.7/site-packages/asl_f4_loader/fw_update.py(92)_send_ping()
-> ep.stop()
  /Applications/Kivy.app/Contents/Resources/venv/lib/python2.7/site-packages/asl_f4_loader/xbvc_py.py(388)stop()
-> self._recv_thread_handle.join()
```

**Comment by brentpicasso** [comment]
Thanks, @mattlqx  - 

question - when you initiate the firmware update, is the app first connected normally to RC before you attempt the firmware upgrade?

Reason why I ask is, the app sends a command to RC to reboot in bootloader mode - but to do that it needs to be connected and communicating normally, first. 

If it does successfully get into bootloader mode the LED will blink green, rapidly.

**Comment by mattlqx** [comment]
Yes, blue and green lights. Then initiate update, and it turns to fast blinking green and stays there after failure until physically disconnected and recycled.

**Comment by brentpicasso** [comment]
@mattlqx 's trace was helpful. While we were not able to repro it here, we made the bootloader client code more tolerant to issues, which should (hopefully) address the problem. https://github.com/autosportlabs/XBVC/commit/811a469a07fcfe64a7d5ffcfee21b806d5083bd7

updated asl_f4_loader package which uses XBVC: https://github.com/autosportlabs/RaceCapture_App/commit/7faa82a13e5a26a9164e031ed2d605238669dd14

This fix should be  available in app 1.12.0
